🚨🚨🚨 [`Refactor`] Move third-party related utility files into `integrations/` folder 🚨🚨🚨

younesbelkada opened the pull request:
# What does this PR do?

As per title and to address https://github.com/huggingface/transformers/pull/25077#discussion_r1283042392 
Let's move all third party libs (outside HF ecosystem) related utility files inside `lib_integrations/`, currently to the best of my knowledge bitsandbytes and deepspeed are the only 2 third party libs that we use as a plugin integration and is outside HF ecosystem

cc @sgugger and @stas00 as it touches DS related code

Let me see first if the CI passes

younesbelkada committed:
move deepspeed to `lib_integrations.deepspeed`

younesbelkada committed:
more refactor

younesbelkada committed:
oops

younesbelkada committed:
fix slow tests

younesbelkada committed:
Fix docs

younesbelkada committed:
fix docs

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25599). All of your documentation changes will be reflected on that endpoint.

stas00 commented:
Hi Younes,

Unless I'm mistaken these would be BC breaking changes.

```
$ PYTHONPATH=src python -c "from transformers.deepspeed import HfDeepSpeedConfig"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'transformers.deepspeed'
```


(In general for deepspeed integration issues please tag @pacman100 who is the current maintainer of the integration.)

stas00 commented:
unrelated - perhaps there could be a neater name than `lib_integrations` - this is awkward IMHO.

Perhaps just `integrations` or just `libs` or `external` - in other words one word would be smoother.

sgugger commented:
Yes I had mentioned integrations as well as a name for the folder. Note that we usually do not guarantee backward compatibility with imports not at the init level (anything not imported in the main init or a subfolder init is considered private), but we can keep a deepspeed module that reimports `HfDeepSpeedConfig` if this line is in a lot of places.

younesbelkada commented:
I did not used `integrations` as there is already a module named `integrations.py` in the same place. Maybe renaming `lib_integrations` to `external` would be better, WDYT?

stas00 commented:
Use a single `integration`? Since once you add the specifics it's a clear `transformers.integration.what`

as in ``transformers.integration.deepspeed`

sgugger commented:
As long as the objects are re-imported, it's completely fine changing `integrations.py` to an `integrations` folder.

younesbelkada committed:
addess feedback

sgugger reviewed (commented):
Not sur we need subfolders for DeepSpeed, bitandbytes and peft, they can jus be modules no?

younesbelkada committed:
address feedback

younesbelkada commented:
Ah yes makes sense, modified it accordingly

younesbelkada commented:
Actually sorry, forgot to do it for peft, will do it now

younesbelkada committed:
final modifs for PEFT

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into move-integrations

younesbelkada committed:
fixup

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25599). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
Merge branch 'main' into move-integrations

younesbelkada committed:
ok now

younesbelkada committed:
Merge branch 'move-integrations' of https://github.com/younesbelkada/transformers into move-integrations

younesbelkada commented:
@sgugger the PR is now ready again ! 🙏 
The failing CI is related with the Hub rate limit issue we have been discussing internally (cc @ydshieh), would like to have a review if possible and I can put the PR on hold until the issue gets solved !

younesbelkada committed:
trigger CI

younesbelkada committed:
trigger CI again

sgugger commented on docs/source/en/main_classes/deepspeed.md:
```diff
@@ -2065,20 +2065,20 @@ In this case you usually need to raise the value of `initial_scale_power`. Setti
 ## Non-Trainer Deepspeed Integration
 
-The [`~deepspeed.HfDeepSpeedConfig`] is used to integrate Deepspeed into the 🤗 Transformers core
+The [`~integrations.deepspeed.HfDeepSpeedConfig`] is used to integrate Deepspeed into the 🤗 Transformers core
```
```suggestion
The [`~integrations.HfDeepSpeedConfig`] is used to integrate Deepspeed into the 🤗 Transformers core
```
I think it's better to import directly from `integrations` since it's an important class. (Here and below)

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -661,7 +661,7 @@ def _load_state_dict_into_meta_model(
 
     if is_quantized:
-        from .utils.bitsandbytes import set_module_quantized_tensor_to_device
+        from .integrations.bitsandbytes import set_module_quantized_tensor_to_device
```
Same here and below.

sgugger reviewed (commented):
Thanks for iterating! Let's import everything directly from `integrations`, but the rest LGTM!

pacman100 reviewed (commented):
Thank you Younes for refactoring the integrations in a separate module 🤗.

As Stas mentioned, this is going to be a breaking change. For example, in `Accelerate` repo line https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/dataclasses.py#L728 needs to be updated. This would also mean updating the DeepSpeed documentation. Else the DeepSpeed integration of the trainer will break. 

Other than that, LGTM!

younesbelkada committed:
Update docs/source/en/main_classes/deepspeed.md

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
import from `integrations`

younesbelkada commented on src/transformers/integrations/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    set_hf_deepspeed_config,
+    unset_hf_deepspeed_config,
+)
+from .integration_utils import (
```
Agreed!

younesbelkada committed:
address feedback

younesbelkada commented:
Thank you all! Feedback should be addressed, @pacman100 I am happy to propose a patch in accelerate after this gets merged, regarding the docs, please see: https://github.com/younesbelkada/transformers/blob/move-integrations/docs/source/en/main_classes/deepspeed.md which has been adapted with the refactor, let me know if there are some other place where the doc should be updated

sgugger commented:
@younesbelkada as said before, we will need to keep the current DeepSpeed module to import back objects for backward compatibility (liek `file_utils.py` does), it's too breaking otherwise.

younesbelkada committed:
revert removal of `deepspeed` module

younesbelkada committed:
revert removal of `deepspeed` module

younesbelkada commented:
Ah makes sense, reverted that and added a comment in the file

sgugger reviewed (commented):
No the deepspeed module should just import objects and raise a deprecation warning (again, look at the `file_utils.py` module.

younesbelkada committed:
fix conflicts

younesbelkada committed:
ooops

younesbelkada committed:
oops

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into move-integrations

younesbelkada committed:
add deprecation warning

younesbelkada committed:
place it on the top

younesbelkada commented on src/transformers/deepspeed.py:
```diff
@@ -12,378 +12,29 @@
+
+warnings.warn(
+    "transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations",
+    DeprecationWarning,
```
Should this be a `FutureWarning` instead? I see most of deprecation warnings in transformers are `FutureWarning`, maybe because `DeprecationWarning` are not raised by default (from my tests locally)?

```suggestion
    FutureWarning,
```

sgugger replied:
Yes, it should be a FutureWarning.

younesbelkada replied:
ok thanks, modified it in d50051a

younesbelkada reviewed (commented):
Thanks and sorry for the multiple iterations, should be all good now, I left one question. However I couldn't find any deprecation warning in `file_utils.py`: https://github.com/huggingface/transformers/blob/main/src/transformers/file_utils.py neither on the main init, I also looked at the places where objects from that file being imported and could not see any depr warning as well

sgugger commented on src/transformers/__init__.py:
```diff
@@ -115,8 +116,6 @@
         "is_tensorboard_available",
         "is_wandb_available",
     ],
-    "lib_integrations": [],
```
Missing an `integrations` key here.

younesbelkada replied:
That key is already defined above: https://github.com/younesbelkada/transformers/blob/move-integrations/src/transformers/__init__.py#L108

sgugger reviewed (commented):
Thanks! Do you know why the diff in the not_doctested file is so big?
As for the deprecation warning, I forgot we didn't put one in `file_utils` (probably cause it would be so breaking).

younesbelkada committed:
put `FutureWarning`

younesbelkada committed:
fix conflicts with not_doctested.txt

younesbelkada committed:
add back `bitsandbytes` module with a depr warning

younesbelkada committed:
fix

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into move-integrations

younesbelkada committed:
fix

pacman100 reviewed (approved):
Thank you @younesbelkada for iterating!

younesbelkada commented:
Thanks everyone for your reviews, regarding the large diff on the doctest file my intuition is that it got large because I added a new line that changed the alphabetical order of the file but I am not sure. (maybe @ydshieh can confirm)
I also realised we might need to keep `utils/bitsandbytes.py` otherwise it would be too breaking, so I reverted it back and did a similar approach than `deepspeed.py`

younesbelkada committed:
fixup

younesbelkada committed:
oops

younesbelkada committed:
fix doctests

younesbelkada committed:
Merge branch 'main' into move-integrations

younesbelkada commented:
The large diff is fixed with https://github.com/huggingface/transformers/pull/25680

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into move-integrations

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into move-integrations

younesbelkada commented:
@sgugger , ran the daily CI (with a reduced number of models tests) thanks to @ydshieh and you can see the results here: https://github.com/huggingface/transformers/actions/runs/5965296109 
compared the report with the report of the same day and I don't see any surprising difference / test failure caused by this PR. Therefore I think we can merge it ! Can you please have a final look ? 🙏 Thanks!

sgugger reviewed (approved):
Thanks for all the work on this!
